qualcommax: add 6.12 kernel support

robimarko opened the pull request:
6.12 support for all qualcommax subtargets.

Do note that I only runtime tested ipq807x, rest are only compile test so testing is very appreciated.

robimarko committed:
qualcommax: add 6.12 as testing kernel

Allow testing 6.12 as the testing kernel.

Link: https://github.com/openwrt/openwrt/pull/18795
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
kernel/qualcommax: Restore kernel files for v6.6

This is an automatically generated commit which aids following Kernel patch
history, as git will see the move and copy as a rename thus defeating the
purpose.

For the original discussion see:
https://lists.openwrt.org/pipermail/openwrt-devel/2023-October/041673.html

Link: https://github.com/openwrt/openwrt/pull/18795
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
kernel/qualcommax: Create kernel files for v6.12 (from v6.6)

This is an automatically generated commit.

When doing `git bisect`, consider `git bisect --skip`.

Link: https://github.com/openwrt/openwrt/pull/18795
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
qualcommax: 6.12: drop upstreamed patches

Drop patches that are already present in kernel 6.12.

Link: https://github.com/openwrt/openwrt/pull/18795
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
qualcommax: 6.12: refresh patches

Refresh 6.12 patches, those that failed automatic refresh were refreshed
manually.

DT bindings patches that failed were dropped as we dont use them in practice.

Link: https://github.com/openwrt/openwrt/pull/18795
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
qualcommax: 6.12: apm: change remove to void

Remove op is now of a void type, update APM.

Link: https://github.com/openwrt/openwrt/pull/18795
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
qualcommax: 6.12: cpr: update for 6.12

Remove op is now of a void type and strlcpy was removed.

Link: https://github.com/openwrt/openwrt/pull/18795
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
qualcommax: 6.12: refresh config

Refresh kernel config by running kernel_menuconfig.

Manually enable CONFIG_HW_RANDOM and CONFIG_CRYPTO_DEV_QCOM_RNG as these
got somehow disabled by the refresh.

Link: https://github.com/openwrt/openwrt/pull/18795
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
qca-nss-dp: update for 6.12 compatibility

Update to current QCA-NSS-DP head for 6.12 compatibility.

Link: https://github.com/openwrt/openwrt/pull/18795
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
qca-ssdk: update for 6.12 compatibility

Update to current SSDK head for 6.12 compatibility.

We must disable warnings for missing declarations and prototypes since
whole SSDK is litered with them and actually fixing them is like doing
whack a mole.

Link: https://github.com/openwrt/openwrt/pull/18795
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
qualcommax: 6.12: pwm: fixup for 6.12

6.12 PWM core introduced a bunch of incompatible changes, namely removal
of manual module owner assignment, complete PWM struct allocation and usage
refactor, etc.

So, update the driver to follow other drivers in 6.12 so it compiles.

Link: https://github.com/openwrt/openwrt/pull/18795
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
qualcommax: 6.12: pwm: use devm_clk_get_enabled()

Using devm_clk_get_enabled() allows simplification of the driver and
dropping of the .remove OP as well.

Link: https://github.com/openwrt/openwrt/pull/18795
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
qualcommax: 6.12: MPD: add missing <linux/of_platform.h>

It seems that kernel cleaned up its kernels so we need to include
<linux/of_platform.h> for of_platform_device_create().

Link: https://github.com/openwrt/openwrt/pull/18795
Signed-off-by: Robert Marko <robimarko@gmail.com>

georgemoussalem commented:
> @georgemoussalem It would be great if you could give IPQ50xx a go.

Let me give it a run tomorrow or Friday. Does that timeline work for you?

robimarko commented:
Sure

georgemoussalem commented:
all systems go for ipq50xx on kernel 6.12!
bootlog: https://pastebin.com/sNrKG0cW

ethernet and wifi are working, no driver probe failures :+1: 

Tested-on: Linksys MR5500

danpawlik commented:
@robimarko for gl.inet b3000 it stuck:
```
[    1.251121] In-situ OAM (IOAM) with IPv6
[    1.254733] NET: Registered PF_PACKET protocol family
[    1.259992] 8021q: 802.1Q VLAN Support v1.8
[    1.316508] cpufreq: cpufreq_online: CPU0: Running at unlisted initial frequency: 799999 KHz, changing to: 800000 KHz
[    1.318854] remoteproc remoteproc0: cd00000.remoteproc is available
[    1.326464] qcom-q6-mpd cd00000.remoteproc: pd-1 node found
[    1.334148] remoteproc remoteproc1: pd-1 is available
[    1.337978] qcom-q6-mpd cd00000.remoteproc: pd-3 node found
[    1.344585] remoteproc remoteproc2: pd-3 is available
[    1.354968] UBI error: cannot open mtd rootfs, error -2
[    1.355244][    1.361612] Waiting for root device /dev/ubiblock0_1...
[   11.368515] mdio_bus 90000.mdio-1:11: deferred probe pending: (reason unknown)
```

full log: https://gist.github.com/danpawlik/979add1e78196592dff11bb1fe8b12e6

georgemoussalem commented:
> @robimarko Thank you for your work. I make a test on gl.inet b3000 (ipq5018) and it stuck:
> 
> ```
> [    1.251121] In-situ OAM (IOAM) with IPv6
> [    1.254733] NET: Registered PF_PACKET protocol family
> [    1.259992] 8021q: 802.1Q VLAN Support v1.8
> [    1.316508] cpufreq: cpufreq_online: CPU0: Running at unlisted initial frequency: 799999 KHz, changing to: 800000 KHz
> [    1.318854] remoteproc remoteproc0: cd00000.remoteproc is available
> [    1.326464] qcom-q6-mpd cd00000.remoteproc: pd-1 node found
> [    1.334148] remoteproc remoteproc1: pd-1 is available
> [    1.337978] qcom-q6-mpd cd00000.remoteproc: pd-3 node found
> [    1.344585] remoteproc remoteproc2: pd-3 is available
> [    1.354968] UBI error: cannot open mtd rootfs, error -2
> [    1.355244][    1.361612] Waiting for root device /dev/ubiblock0_1...
> [   11.368515] mdio_bus 90000.mdio-1:11: deferred probe pending: (reason unknown)
> ```
> 
> full log: https://gist.github.com/danpawlik/979add1e78196592dff11bb1fe8b12e6

hmm.. you must have a device with this winbond chip which also supports a minimum of 1 bit ecc strength:

```
SPINAND_INFO("W25N01GWZEIG",
		     SPINAND_ID(SPINAND_READID_METHOD_OPCODE_DUMMY, 0xba, 0x21),
		     NAND_MEMORG(1, 2048, 64, 64, 1024, 20, 1, 1, 1),
		     NAND_ECCREQ(1, 512),  <-- See here: first argument
```

We see this on the AX6000 as well. @TheRootED24: if my memory serves me well, didn't you have a chip which needed 4 bits ECC strength? If so, you could set `nand-ecc-strength = <4>;` in the dts. But that will cause issues if you have chips that require 8 bits ecc strength.

Anyways, the nand chips qcom and vendors have selected for ipq50xx boards all support BCH based on 4 or 8 bits ECC strength and U-boot overrides ecc strength and sets it to 4 if its requirements are less. The old qcom_nandc driver uses more ecc bytes if ecc strength is >= 8, else less:
https://github.com/torvalds/linux/blob/088d13246a4672bc03aec664675138e3f5bff68c/drivers/mtd/nand/raw/qcom_nandc.c#L1400 

@robimarko: we need to fix this as it's now a wider issue for ipq50xx, and perhaps ipq9574 as it uses the spi-qpic-snand driver too.

georgemoussalem commented:
@danpawlik: I've created an experimental patch, can you check if that works for you?
https://filebin.net/evumslku539w745m

you won't have to set the nand-ecc-strength property in the dts.

danpawlik commented:
Woo, nice work @georgemoussalem ! It works \o/
I see wireless network is available, but can not make test right now.

georgemoussalem commented:
> @georgemoussalem Cant we simply patch the driver for now to warn is 1-bit ECC is set and set 4-bit ECC instead?

Yes, that would work. Will create a patch

mrnuke commented:
Full steam ahead on **ipq60xx**::linksys-mr7350. I see no apparent issues with the 6.12 kernel.

robimarko committed:
qualcommax: ipq50xx: spi-qpic-snand: default to 4-bit ECC

There are NAND IC-s that define 1-bit ECC as the minimal strength,
however that is unsupported by QPIC-SNAND as it only supports 4 or 8 bit
ECC.

Since most of these chips also support 4-bit ECC just fine, instead of
erroring out if 1-bit ECC is requested lets instead default to 4-bit ECC.

Fixes: 01b72ce61e8f ("qualcommax: ipq50xx: remove ECC user config from board files")
Signed-off-by: George Moussalem <george.moussalem@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18795
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
@danpawlik It should work now on your board aswell

TheRootED24 commented:
@danpawlik  did you get a chance to test ? 

I'll do a quick build and confirm

danpawlik commented:
@robimarko thank you - it works \o/ All is working perfect (after few hours).

georgemoussalem commented on target/linux/qualcommax/patches-6.12/0162-arm64-dts-qcom-ipq5018-Add-PCIe-related-nodes.patch:
The pcie phy nodes are out of place, they should come right after the usb phy.

robimarko replied:
I took the patch as-is from the v9 upstream submission, its possible that quilt refreshed it this way due to patch ordering we have downstream.
